src/main/java/com/revlearn/team1/config/JwtAuthenticationFilter.java

```diff
@@ -0,0 +1,75 @@
+package com.revlearn.team1.config;
+
+import com.revlearn.team1.exceptions.UserNotFoundException;
+import com.revlearn.team1.model.User;
+import com.revlearn.team1.service.user.UserService;
+import com.revlearn.team1.util.JwtUtil;
+import jakarta.servlet.FilterChain;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
+import org.springframework.security.web.util.matcher.AntPathRequestMatcher;
+import org.springframework.stereotype.Component;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+@Component
+@RequiredArgsConstructor
+public class JwtAuthenticationFilter extends OncePerRequestFilter {
+    private final UserDetailsService userDetailsService;
+    private final JwtUtil jwtUtil;
+    private final UserService userService;
+
+    @Override
+    protected boolean shouldNotFilter(@NonNull HttpServletRequest request) throws ServletException {
+        final List<String> skipFilterUrls = Arrays.asList("/auth/**");
+        return skipFilterUrls.stream().anyMatch(url -> new AntPathRequestMatcher(url).matches(request));
+    }
+
+    @Override
+    protected void doFilterInternal(@NonNull HttpServletRequest request, @NonNull HttpServletResponse response, @NonNull FilterChain filterChain) throws ServletException, IOException {
+        final String authHeader = request.getHeader("Authorization");
+        final String jwt;
+        final Long userId;
+
+        if (authHeader == null || !authHeader.startsWith("Bearer ")) {
+            filterChain.doFilter(request, response);
+            System.out.println("No auth header");
+            return;
+        }
+
+        jwt = authHeader.substring(7);
+
+        userId = jwtUtil.extractUserId(jwt);
+        User user = userService.findById(userId.intValue())
+                .orElseThrow(() -> new UserNotFoundException(String.format("User with id %d not found", userId)));
+        System.out.println("User found: ");
+        if (SecurityContextHolder.getContext().getAuthentication() == null) {
+            System.out.println("Authenticating user");
```

Comment:
also you don't have to do this for this PR (feel free to auto resolve this comment) but if you wanna start replacing the System.out.printlns with the logger.info and logger.error that daniel implemented, that'd be neato :)